[RFC] Introduction to MSPI (Multi-bit SPI) driver API #70723

swift-tk opened the pull request:
### Description
This PR primarily forces on creating a generic MSPI(multi-bit SPI) controller API that typically is used when interfacing with high speed pSRAM/flash memory, displays, and sensors devices.

I have finished below items from https://github.com/zephyrproject-rtos/zephyr/issues/70723#issuecomment-2045544160. Also see this issue for more detailed RFC.

- MSPI Header file
- MSPI emulator under bus emulators
- A bus API test that runs under the MSPI emulator
- Documentation
- At least one implementation of the new bus API (i.e. the hardware peripheral)
- A device driver that uses the new bus API (e.g. a flash memory)

### Q & A
#### Q: Why include sensors?
A: The MSPI would still be compatible with generic serial SPI interface and the API is expandable to support RTIO.
	
	
#### Q: Use of Kconfig MSPI_COMPLETION_TIMEOUT_TOLERANCE
A: This is a maximum time allowed for a transfer regardless of error or large data size. 
   The unit is not yet determined, maybe in milli-seconds.

#### Q: Why won't we include a parameter for the maximum time to wait for the next transfer.
A: MSPI only cares about the transfer, it is all about the transfer. It does not care what happens in between transfers. 
   Unlike SDIO, the protocol dictates a controlled flow and that there is a response for every command. 
   This is not true for SPI. The MSPI controller can't be responsible for knowing the internal state of specific devices. 
   This is the responsibility of the specific device driver.
   
#### Q: Purpose of `struct mspi_dev_cfg`
A: This contains the device specific hardware controller settings. However, they are device properties and should be derived from device datasheets.  While often time the value of its member may frequently change, the settings derived from DTS that filled within the struct should represent the most frequently used settings during run-time.
When value of its member changes, the mspi_dev_config API should be used for device operation mode switching.
   
For future enhancement, enhanced XIP or DMA that requires sending only cmd and address one time may be supported but will not be in the context of this PR.

#### Q: Support for memory mapped device
A: The XIP address should be obtained from devicetree memory regions. For multi-slave scenario, each device's mapped address may be determined by the start address along with device index and its size. This can be done within the controller or pass by `struct mspi_xip_config`

Going to defer code linking from this PR as it is more of kernel enhancement and also for the fact that we may need more time to think about it.
   
#### Q: Reasons when choosing the member of `struct mspi_xfer`
A: 
- The MSPI API is lower level controller API and there should be minimum overlap between what MSPI APIs could do.
- For mspi_transceive, there should only be transfer related settings that has to do with cmd, address, data, and dummy phases.
- By considering real world usage of SPI devices and balancing between different controllers and devices, it is not worth to include parameters that describe the device's operation modes such as number of lines, frequency and other parameters that is within `struct mspi_dev_cfg`. For the fact that, devices may require different ways to manage operation modes and controller hardware/hal implementation differences. The purpose of any high speed MSPI implementation should be data streaming as fast as possible so it doesn't seem necessary to include mode switching in this level of API while software queue like RTIO might be more appropriate when 90% of the time during run-time should require no mode switching.
	  
#### Q: MSPI slave usage
A: When an MSPI controller act as a slave, it is essentially an MSPI device. A slave device driver could be created for this purpose to interface with the existing controller driver and the APIs would still apply. But there is no plan to show an implementation of a mspi slave in this PR and the details may subject to change.

If you think there might be a conflict with the current proposal, please point it out.
At this point, the use of MSPI slave may be very specific and uncommon.

### Action items:

- [x] 	To add an additional timeout parameter in `struct mspi_xfer` to configure a time out value per transfer, this value should be smaller than CONFIG_MSPI_COMPLETION_TIMEOUT_TOLERANCE. Left the timeout as constant in device drivers for the time being.
- [x]  To expand `mspi_data_rate` as devices may switch between 1s-4s-4s, 1s-4s-4d and 1s-4d-4d, and all other possible combinations. It now has four types: S_S_S, S_S_D, S_D_D, D_D_D. In combination with `mspi_io_mode`, more transfer types can be supported. 
- [ ]  Would like to consolidate on API return values
   - `mspi_ambiq_config`. 
      - 0 success.
      - `-ENOTSUP` for when member of `mspi_cfg` value is not supported by the controller.
      - `-EHOSTDOWN` for HAL return fail.
      - `-EPERM` for fail to initialize because the controller has already been initialized.
    - `mspi_dev_config` 
      - 0 success
      - `-EBUSY` for when the controller is busy due to in transfer or gaining access mutex time out.
      - `-ENODEV` for failing to validate `mspi_dev_id`.
      - `-ENOTSUP` for when member of `mspi_dev_cfg` or `mspi_dev_cfg_mask` not supported by the controller.
      - `-EHOSTDOWN` for HAL return fail.
    - `mspi_xip_config`
       - 0 success
       - `-ESTALE` for when the request is denied as it doesn't own access mutex.
       - `-EHOSTDOWN` for HAL return fail.
       - `-EBUSY` for when the controller is still busy processing transfers and does not allow configuration.
    - `mspi_scramble_config`
       - 0 success
       - `-ESTALE` for when the request is denied as it doesn't own access mutex.
       - `-EHOSTDOWN` for HAL return fail.
       - `-EBUSY` for when the controller is still busy processing transfers and does not allow configuration.
    - `mspi_timing_config`
       - 0 success
       - `-ESTALE` for when the request is denied as it doesn't own access mutex.
       - `-EHOSTDOWN` for HAL return fail.
       - `-EBUSY` for when the controller is still busy processing transfers and does not allow configuration.
       - `-ENOTSUP` for when the parameters not supported by the controller.
    - `mspi_get_channel_status`
       - 0 success
       -  `-EBUSY` for when the controller is still busy processing transfer
       - Need more return values such as XIP status
    - `mspi_transceive`
       - 0 success
       - `-ESTALE` for when the request is denied as it doesn't own access mutex.
       - `-EIO` for when IO error occurs
       - `-EFAULT` for incorrect `mspi_xfer` parameters such as empty buffer
       - `-EBUSY` for timeout when gaining the the transfer context lock. 
       - `-ENOMEM` for when the hardware command queue is full.
   
- [ ] Consider how to properly link code/data from MSPI XIP region, especially in multi-slave scenario. This https://github.com/zephyrproject-rtos/zephyr/pull/68888 can work, however, there might be some unknown side effects when moving to PRE-KERNEL.

swift-tk committed:
drivers: mspi: Add the new MSPI API

The MSPI(multi-bit SPI) is provided as a generic API to accommodate
advanced SPI peripherals and devices that typically require command,
address and data phases, and multiple signal lines during these phases.
While the API supports advanced features such as XIP eXecute In Place
and scrambling, it is also compatible with generic SPI.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
dts: mspi: Add MSPI bindings

Add the generic controller and device bindings for MSPI.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
doc: Add MSPI to bus emulator documentation

Add the MSPI bus emulator to documentation.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
doc: Add MSPI API documentation

Add the mspi.rst doc and related things for document generation.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
dts: mspi: Add MSPI emulator bindings

Add the controller and device emulator bindings for MSPI.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
boards: dts: Add MSPI controller to native_sim

Add the MSPI emulator controller node in the board dts.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
emul: mspi: Add the mspi controller emulator

Add bus emulator support for MSPI and the MSPI controller emulator.
The mspi_emul.c not only serves as an emulator but also provides an
example implementation of the MSPI API. It does not actually do anything
other than validating parameters and forwarding transceive request back
to the device driver emulators.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
tests: mspi: Add mspi api test

Add a MSPI controller API test and run on native_sim board.
The API implementation within mspi_emul.c is the test target.
The stub device emulator is added for convenience to test controllers.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
dts: mtd: Add MSPI flash emulator binding

Add the binding the flash emulator under MSPI bus.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
drivers: flash: Add a flash emulator

Add a flash emulator that uses MSPI API to emulate a flash device
on MSPI bus.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
tests: mspi: Add a mspi flash test

The mspi flash test is intended to test mspi flash device driver
implementation on MSPI bus. The test requires the device to have at
least two sectors of memory with MSPI_FLASH_SECTOR_SIZE as size.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
soc: ambiq: Add shared_ram section

Add linker script file shared_ram.ld where defines sections used by
specific peripherals.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
dts: mspi: Add Ambiq MSPI DTS and bindings

Add the Ambiq MSPI nodes to soc device tree and base bindings for
MSPI controllers and devices.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk commented:
Hi @carlescufi 
Could you please add reviewers to the PR and possibly link to https://github.com/zephyrproject-rtos/zephyr/issues/70723. Thanks.

swift-tk committed:
drivers: mspi: Add Ambiq apollo3p mspi controller

The Ambiq MSPI controller is implemented using the MSPI bus API.
The hardware supports up to 48MHz octal SDR with XIP, scrambling and
hardware command queue features.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
drivers: memc: Add APS6404L device driver

The APS6404L psram is a quad SDR SPI device that runs up to 100MHz.
It can provide 8MB of external RAM for SoCs that supports XIP feature.
The device driver uses MSPI bus API and could be used across
different controllers that implement the MSPI bus API.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
samples: memc: Add APS6404L device to memc example

Demo the usage of APS6404L device driver in memc example.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
drivers: mspi: Add ATXP032 NOR flash driver

The ATXP032 is a NOR flash device that supports up to ~100MHz
octal SDR/DDR with 4MB nonvolatile memory.
The device driver uses MSPI bus API and could be used across different
SoC controllers that implement the MSPI bus API.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
tests: mspi: Add apollo3p_evb board to flash test

The apollo3p_evb board has a ATXP032 NOR flash and driver is tested
in mspi/flash test.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

kartben commented on doc/hardware/emulator/bus_emulators.rst:
```diff
@@ -88,7 +88,7 @@ The ``EMUL_DT_DEFINE()`` function accepts two API types:
   #. ``bus_api`` - This points to the API for the upstream bus that the emulator
      connects to. The ``bus_api`` parameter is required.  The supported
-     emulated bus types include I2C, SPI, and eSPI.
+     emulated bus types include I2C, SPI, eSPI and MSPI.
```
nit :)

```suggestion
     emulated bus types include I2C, SPI, eSPI, and MSPI.
```

kartben commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,197 @@
+.. _mspi_api:
+
+Multi-bit SPI Bus
+###########################################
```
```suggestion
Multi-bit SPI Bus
#################
```

kartben commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,197 @@
+Multi-bit SPI Bus
+###########################################
+
+The MSPI(multi-bit SPI) is provided as a generic API to accommodate
```
```suggestion
The MSPI (multi-bit SPI) is provided as a generic API to accommodate
```

kartben commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,197 @@
+The MSPI(multi-bit SPI) is provided as a generic API to accommodate
+advanced SPI peripherals and devices that typically require command,
+address and data phases, and multiple signal lines during these phases.
+While the API supports advanced features such as XIP eXecute In Place
```
```suggestion
While the API supports advanced features such as XIP eXecute In Place
```
```suggestion
While the API supports advanced features such as :term:`XIP`
```

kartben commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,197 @@
+*******************
+
+Zephyr's MSPI controller API may be used when a multi-bit SPI controller
+is present. E.g. Ambiq MSPI, QSPI, OSPI, Flexspi and etc.
```
```suggestion
is present, e.g. Ambiq MSPI, QSPI, OSPI, Flexspi, etc.
```

kartben commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,197 @@
+Zephyr's MSPI controller API may be used when a multi-bit SPI controller
+is present. E.g. Ambiq MSPI, QSPI, OSPI, Flexspi and etc.
+The API supports single to hex SDR/DDR IO with variable latency and advanced
+features such as XIP and scrambling. Applicable devices includes but not
+limited to high-speed, high density flash/psram memory devices, displays
```
```suggestion
features such as XIP and scrambling. Applicable devices include, but not
limited to, high-speed, high density flash/psram memory devices, displays
```

kartben commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,197 @@
+
+The MSPI interface contains controller drivers that are SoC platform specific
+and implement the MSPI APIs, and device drivers that reference these APIs.
+The relationship between the controller and device drivers is many to many to
```
```suggestion
The relationship between the controller and device drivers is many-to-many to
```

kartben commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,197 @@
+
+Transceive
+===========
+The transceive request is of type :c:struct:`mspi_xfer` which allows for dynamically
+changing the transfer related settings once the mode of operation is determined and
+configured by :c:func:`mspi_dev_config`.
+The API supports bulk transfers with different starting addresses and sizes with
+:c:struct:`mspi_xfer_packet`. However, it is up to the controller implementation
+whether to support scatter IO and callback management. The controller can determine
+which user callback to trigger based on :c:enum:`mspi_bus_event_cb_mask` upon completion
+of each async/sync transfer if the callback had been registered using
+:c:func:`mspi_register_callback`. Or not to trigger any callback at all with
+:c:enum:`MSPI_BUS_NO_CB` even if the callbacks are already registered.
+In which case that a controller supports hardware command queue, user could take the full
+advantage of it in terms of performance if scatter IO and callback management are supported.
```
you might want to add blank lines to actually create separate paragraphs

kartben commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,197 @@
+Device Tree
+===========
+
+Here is an example for defining a MSPI controller in device tree:
```
```suggestion
Here is an example for defining an MSPI controller in device tree:
```

swift-tk replied:
I thought we are just supposed to use "an" before vowels. Are you sure?

kartben replied:
"an" is used before a word that starts with a vowel *sound*, not necessarily just a vowel letter, in this case: "an emm-SPI", just like you would write "an SPI bus", "an FBI agent", etc.

swift-tk replied:
Yes, thank you.

kartben commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,197 @@
+mspi-io-mode and mspi-data-rate even though they may subject to change during run-time.
+Typically it should represent the condition of the device during normal operations.
+
+Muti-slave
+==============================
```
```suggestion
Muti-slave
==========
```

kartben commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,197 @@
+* :kconfig:option:`CONFIG_MSPI_SCRAMBLE`
+* :kconfig:option:`CONFIG_MSPI_TIMING`
+* :kconfig:option:`CONFIG_MSPI_INIT_PRIORITY`
+* :kconfig:option:`MSPI_COMPLETION_TIMEOUT_TOLERANCE`
```
```suggestion
* :kconfig:option:`CONFIG_MSPI_COMPLETION_TIMEOUT_TOLERANCE`
```

kartben reviewed (changes requested):
will make another pass on the docs when other folks actually knowledgeable about the topic will have provided their own reviews. It is looking pretty good and thorough though, so thanks for that!

swift-tk commented:
> will make another pass on the docs when other folks actually knowledgeable about the topic will have provided their own reviews. It is looking pretty good and thorough though, so thanks for that!

@kartben  Many thanks!

swift-tk committed:
samples: mspi: Add a mspi flash sample

The sample code is copied from spi_flash. To flash API, there is no
difference if the bus is spi or mspi.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
samples: mspi: Add a mspi async sample

The sample is added to demonstrate mspi_transceive asynchronously.
Apollo3 MSPI controller has hardware command queue and supports
scatter IO and callback management, so no additional software queue
is required.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

erwango commented on include/zephyr/drivers/mspi.h:
```diff
@@ -0,0 +1,791 @@
+/**
+ * @brief MSPI transfer directions
+ */
+enum eTransDirection {
```
```suggestion
enum trans_direction {
```

erwango commented on include/zephyr/drivers/mspi.h:
```diff
@@ -0,0 +1,791 @@
+ */
+struct mspi_cfg {
+	/** @brief mspi channel number */
+	uint8_t                 ui8MSPIChannel;
+	/** @brief Configure operaton mode */
+	enum mspi_op_mode       eOPMode;
+	/** @brief Configure duplex mode */
+	enum mspi_duplex        eDuplex;
+	/** @brief DQS support flag */
+	bool                    bDQS;
+	/** @brief Software managed multi-slave enable */
+	bool                    bSwMultiSlave;
+	/** @brief GPIO chip-select line (optional) */
+	struct gpio_dt_spec     *pCE;
+	/** @brief GPIO chip-select line numbers (optional) */
+	uint32_t                ui32CeGpioNum;
+	/** @brief Slave number from 0 to host controller slave limit. */
+	uint32_t                ui32SlaveNum;
+	/** @brief Maximum supported frequency in MHz */
+	uint32_t                ui32MaxFreq;
+	/** @brief Whether to re-initialize controller */
+	bool                    bReinit;
```
As per zephyr coding guidelines, snake_case is required.
I don't think there are explicit requirements about the fact of including the variable type in the name, but I haven't seen this elsewhere in zephyr, so I'd suggest to drop.

swift-tk replied:
Do you happen to have link to the zephyr  doc saying snake case must be used? I couldn’t find it anywhere and it was why I kept it till now.

erwango replied:
I guess the assumption is that this is covered by Linux coding guidelines mentioned in https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style

erwango commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,198 @@
+is present. E.g. Ambiq MSPI, QSPI, OSPI, Flexspi, etc.
+The API supports single to hex SDR/DDR IO with variable latency and advanced
+features such as :term:`XIP` and scrambling. Applicable devices include but
+not limited to high-speed, high density flash/psram memory devices, displays
+and sensors.
```
Do we really want to expand to sensors ?

erwango replied:
ok, this was not my understanding while it was discussed (in order to distinguish with SPI API), but if this is fine with everyone, so be it.

teburd replied:
I'd note that without supplying an ISR safe bus implementation that works as RTIO expects, sensor drivers in Zephyr will not work with this bus.

For sensors, any bus implementation will need to accept and start asynchronous bus transactions from an ISR. Otherwise we are back to dealing with threads/work queues per sensor which gets ridiculous quickly.

teburd replied:
> Sure, for sensors that have cmd address and data phases even though they may not be multi-bit. The API is also compatible with RTIO if the controller is implemented in a certain way.

I don't see any form of submit API here so no I don't think that's true, and I don't think this API will work with the direction sensors are going today without changes.

swift-tk replied:
> > Sure, for sensors that have cmd address and data phases even though they may not be multi-bit. The API is also compatible with RTIO if the controller is implemented in a certain way.
> 
> I don't see any form of submit API here so no I don't think that's true, and I don't think this API will work with the direction sensors are going today without changes.

The documentation saying that the API supports the communication interface required by these categories of devices. The sensors here don’t necessarily means the sensor subsystem.

You are correct, an additional API is needed to support RTIO. In addition, the compatibility will also depend on the controller implementation. 

I have no plan to add that in this PR as we don’t have a sensor to test that. But if we do, the RTIO path will be followed.

danieldegrasse commented on drivers/mspi/Kconfig:
```diff
@@ -0,0 +1,64 @@
+	help
+	  Device driver initialization priority.
+
+config MSPI_COMPLETION_TIMEOUT_TOLERANCE
```
Is this intended to be a command timeout parameter? This seems like something we should pass into an API call rather than implementing it at this level- different commands will have different timeout tolerances.

swift-tk replied:
I agree, I think I just copied that one from SPI and I haven’t given much thought on that one as the devices I work with does not timeout on command but for the entire transfer. I’m considering add that as a param in `mspi_xfer_t`.  However a maximum timeout for the controller itself may still be meaningful.

danieldegrasse replied:
Personally, I think transfer timeouts make more sense. While most devices will likely have reasonable timeouts, I can imagine something like a chip erase command having a very long timeout needed.

It's also worth noting that in my experience a lot of commands won't have a timeout per se- rather you'd send a command to the device, than poll some type of status bit until the flash indicated the operation was complete.

What types of devices have you encountered where a transfer timeout would be meaningful? I'm just interested to know- not opposed to having the functionality in the API

RichardSWheatley replied:
This is the ATXP Series and from what I've seen about Multi-but SPI devices, there are a few that have these requirements.

So in each case, there is a different timing requirement.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/12254017/514aeefa-be77-459f-bad3-9589486752be)

RichardSWheatley replied:
Why would the timeout for a specific command not be passed to the MSPI-controller?

Just a question.

swift-tk replied:
> Why would the timeout for a specific command not be passed to the MSPI-controller?
> 
> Just a question.

Because MSPI only cares about the transfer, it is all about the transfer. It does not care what happens in between transfers. Unlike SDIO, the protocal dictates a controlled flow and that there is a response for every command. This is not true for SPI. The command timeout is not really the correct term here. I think what you are looking for is the minimum time to wait to start the next transfer. However, the MSPI controller can't be responsible for knowing the internal state of specific devices. This is the responsibility of the specific device driver.

danieldegrasse replied:
> The transfer timeout is mainly for when there is an error occurred. The error could be caused by a misconfiguration or that the device did not receive correct command/addr/data due to timing issue, or even the device did not respond at all due to unknown issue. Surely we wouldn't want to wait forever

I agree, waiting forever does not make sense. However, as you pointed out a lot (if not all) of the device wait times are something the device driver would likely manage. In my experience with MSPI commands, the read phase (where I would expect a timeout to occur) is just handled by the controller sending additional clocks to the device. The *controller* has no idea what the data the device returns here means, only the device driver for the specific attached MSPI device would know this.

What I'm trying to understand is essentially the following- what case exists where we would need the *controller* to return an error due to a timeout? I can absolutely see the controller returning an error due to a transfer failing or not being supported, but these errors will occur at configuration time. I'm trying to understand when the controller itself would timeout. If the device requests a long transfer via the `mspi_transfer` function, the controller should just block until that transfer completes, right?

swift-tk replied:
The timeout would also apply to the maximum transfer time as well even if there is no error for the same reason that the upper layer can't wait forever. Thats what the Kconfig `MSPI_COMPLETION_TIMEOUT_TOLERANCE` is for.

swift-tk replied:
> Ok, so if this value was set to 200ms and a driver issued a command that (for some reason) took more than 200 ms to transfer the controller would report an error? I guess the idea is to set an upper time bound on the transfer, but wouldn't it make more sense for _each transfer_ to have an upper time bound, not for the framework to have an upper bound set via Kconfig?

There is a timeout parameter in struct mspi_xfer represents the upper bond timeout of a transfer. The Kconfig here is the one for controllers or Hal as I saw timeout macro defined in vendors’ HAL. The idea is the this number should be the biggest timeout of all and the controller should check the timeout in mspi_xfer against this number so that the program wouldn’t wait further after the controller had given up.

danieldegrasse commented on include/zephyr/drivers/mspi.h:
```diff
@@ -0,0 +1,791 @@
+	/** @brief Configure number of clock cycles between
+	 * addr and data in RX direction
+	 */
+	uint32_t                ui32RXDummy;
```
I'm unclear why these values are configured once per device- different instructions may require different numbers of dummy cycles, right?

Also, what is the intended use of the Write instruction and Read Instruction in this structure? Are these intended to enable memory mapped access to the MSPI device?

RichardSWheatley replied:
Have you looked at all of the code in context?
mspi_xfer_config specifically?

That explicitly implies that you can reconfigure certain things for each transfer. Since each MSPI controller will implement those differently, it's up to the chip manufacturer to add what the wish to add.

Items in mspi.h can be used either in the controller or in the device.

danieldegrasse replied:
One other question- unless I've missed it, I don't see a method for supporting enhanced XIP (IE no command code needed, simply send the address). Perhaps we should add a boolean flag to indicate to the controller that enhanced XIP mode is desired, and a way for the controller to indicate support for this?

swift-tk replied:
The write and read instructions in `mspi_dev_config` are device properties. They represent the most frequent used transfer commands throughout run-time. And yes, they are also for configuring XIP but falls into the device specific controller settings.

As for enhanced XIP or DMA, I don’t have devices and controllers that supports this feature and I’m unfamiliar with the flow. Maybe it is better to defer this from this PR, if it does not introduces conflict with the current proposal. I’d imagine it be added to both mspi_config and mspi_dev_config.

danieldegrasse replied:
Personally, I think we at least need to consider what XIP/memory mapped access looks like in this PR (although we don't have to implement it here). I want to avoid needing major API changes to implement XIP support.

At least on the NXP FLEXSPI, memory mapped access looks like so:
- the FLEXSPI has a programmable lookup table which stores command descriptors for various flash operations (IE send command, send address, read data, etc). The FLEXSPI can be instructed to manually execute any of these command descriptors via register access
- alternatively for memory mapped access, the FLEXSPI can have a READ and WRITE command index programmed for each device (up to 4 attached devices are supported, and they will be memory mapped sequentially in the FLEXSPI address space based on the chip select port used)
- The READ and WRITE command index refers to a specific index in the lookup table. The FLEXSPI can run multiple instructions for one READ or WRITE, since it executes commands until it encounters a special STOP command

To summarize, I think the current `mspi_dev_config` would work well for memory mapped access with the FLEXSPI. The only API I see missing is a way to get the address of a memory mapped device, so that the program can use that address as flash or RAM. We might want a way to read that address statically, since I can imagine users wanting to link data to a PSRAM region or something similar.

swift-tk replied:
I do find it curious that you had to implement a mspi_xip_get_address for the fact that the physical XIP start address is fixed for each of the hardware instance. What I did is to write the start address and size down in devicetree and get it via DT macros.  Also to properly link code, you'd have the XIP memory region defined.
<img width="537" alt="xip_region" src="https://github.com/zephyrproject-rtos/zephyr/assets/155429447/d7f3c089-dba0-4a2f-b912-e05a43030d0f">

danieldegrasse replied:
> I do find it curious that you had to implement a mspi_xip_get_address for the fact that the physical XIP start address is fixed for each of the hardware instance. What I did was to write the start address and size down in devicetree and get it via DT macros. Also to properly link code, you'd have the XIP memory region defined.

Yes, I agree. I think this should probably be moved to devicetree in this API, so that code can be linked to these regions. It's also worth noting that once multiple devices are attached to the FLEXSPI, the address region of each is no longer fixed. Once all devices are configured, their address regions will be sequential (devices would use individual mode here):

![image](https://github.com/zephyrproject-rtos/zephyr/assets/34665051/d30bba77-11d3-499b-b2fd-6d0a289116a0)

However, if FLASH_A1 was not yet configured, the address space of FLASH_A2 would start at the beginning of the XIP region until FLASH_A1 was initialized. At runtime, this rarely makes a difference as all flash devices will be set up. However, it does affect things during init time. If we want to link code into these regions, we will need to implement some sort of function similar to `z_data_copy`: https://github.com/zephyrproject-rtos/zephyr/blob/23a03b7a0a494e27316fdd7fdb453a0409b99aa9/kernel/xip.c#L26 that runs after all MSPI memory mapped devices are initialized. Just something worth considering if we want complete XIP support

swift-tk replied:
Couldn't you determine the address location of each slave depending on its size and device index (from `struct mspi_dev_id`) and configure accordingly? If FLASH_A2 can not get to the correct address without FLASH_A1 being initialized, this seems rather couter intuitive. 
I agree that an general MSPI XIP region binding and corresponding support to move the data are needed. But lets think about it and I prefer to defer it to future enhancement, as it may not be a part of the API but a kernel enhancement.

danieldegrasse replied:
> If FLASH_A2 can not get to the correct address without FLASH_A1 being initialized, this seems rather couter intuitive

You'll have to take that up with NXP's hardware designers :). Essentially, if the size of flash A1 is set within the FLSHA1CR0 register, then the base address of flash A2 will be shifted by the value written to FLSHA1CR0 * 1024. We could theoretically write code for the FLEXSPI to initialize all FLSHAxCR0 registers at boot time, before each individual flash device initialized. However, this wouldn't add much value since the flash devices still couldn't be accessed until they were initialized.

Anyways, no reason to get into a tangent here- as you pointed out, this will likely need to be a future enhancement. I think the main thing we need at this stage is a way to read the base address for each MSPI device at build time, so that can be integrated with the linker and kernel in the future.

swift-tk replied:
I saw that the flexspi XIP start address and size has already been written to nxp soc dts.
```
flexspi: spi@134000 {
	reg = <0x40134000 0x1000>, <0x08000000 DT_SIZE_M(128)>;
};
```
But there is no xip regions. Maybe you can create partitions in a XIP region for a flexspi Or individual memory region to represent each device memory mapped area, since the device size and index are fixed during run-time. However we would not be able to use zephyr_code_relocate as `z_data_copy` is invoked before any device initialization.

Another way is to create partition under controller node, seems to related to https://github.com/zephyrproject-rtos/zephyr/issues/34167.

erwango replied:
> However we would not be able to use zephyr_code_relocate as z_data_copy is invoked before any device initialization.

See https://github.com/zephyrproject-rtos/zephyr/pull/68888 for a related problem on delayed init for relocated data. Tbc isf similar approach can be used for code relocation.

danieldegrasse commented on include/zephyr/drivers/mspi.h:
```diff
@@ -0,0 +1,791 @@
+	/** @brief Configure frequency */
+	uint32_t                ui32Freq;
+	/** @brief Configure I/O mode */
+	enum mspi_io_mode       eIOMode;
```
I'm not sure I understand the idea of the IO mode parameter- many devices have some commands that support quad pad, while other commands only support single pad I/O. This feels like something we should be configuring at the transfer level?

swift-tk replied:
The settings in `mspi_dev_config` typically represent the most frequently used configurations required by the device. I know that some devices does not require explicit op-mode as a device current state but distinguishes the line modes only commands. But in real world usage, only one specific op-mode would be used for most of transfer operations for optimal performance. So it is not so different with devices that enters and exits certain op-mode by command. 

The reason why I split the configurations into mspi_dev_config and mspi_xfer is because switching between op-mode may require quite some time for both device and controller. To optimize performance, mspi_dev_config is dedicated for configuring the controller based on device requirements. Only a minimum device specific settings that may vary per transfer is included in mspi_xfer.

danieldegrasse replied:
My concern here is that the IO mode for read might be different than the IO mode for write. Perhaps the controller supports 1-8-8 read mode at the selected frequency, but only 1-4-4 write mode. I'm not familiar with the Ambiq controller- is there significant overhead to switch operating modes when using it? If so, how would something like the hypothetical device I describe above function?

swift-tk replied:
I'm aware that these type of device do exist, however not in the way you described but does involves ddr/sdr switching. I'm also considering combining the `mspi_op_mode` and `mspi_data_rate`.
These types of device should call `mspi_dev_config` for op mode switching and then call `mspi_transceive`.

The io mode switching in general should just be writing to controller registers. However, there are a number of factors we need to consider.
1. Does the device require mode switching also? If it does, would it require some time?
2. Does the controller have a hardware command queue, does it support changing all these settings and does the controller layer have access to it through HAL? 
   - If all of the above is true for everybody, then yes we can add it `mspi_xfer`. But I don't think this is the case.
   - If without a hardware command queue or without complete access to it, accessing controller configuration would be slower and you may have to configure the hardware every time when calling `mspi_transceive`. This is also what I really dislike the current SPI.  As we have so many layers of HAL API calls and Zephyr API calls. The hardware performance would get reduced very quickly.
3. The MSPI API is a lower API in the hierarchies of all Zephyr APIs. Thus having clear boundaries of each MSPI API should do is always better.
   - There is RTIO that act as a general work queue API people could use for chaining different MSPI API calls if the device operation mode switching become too tedious to handle.

danieldegrasse commented on include/zephyr/drivers/mspi.h:
```diff
@@ -0,0 +1,791 @@
+/**
+ * @brief MSPI peripheral xfer packet format
+ */
+struct mspi_xfer_packet {
```
I believe each packet should support setting the number of pads to use- this will allow us to better support cases where a MSPI device might support 1-4-4 reads, for example

swift-tk replied:
As I indicated in other comment, switching between op-modes may require quite some time for both controller and devices. The mspi_xfer_packet is intended for bulk transfer and scatter IO operation not combining register access and memory access which may have different op-mode requirements.

danieldegrasse replied:
> As I indicated in other comment, switching between op-modes may require quite some time for both controller and devices

But what would the disadvantage be of combining a register and memory access command into one `mspi_transceive` call? If the controller hardware knows it will be faster to "chunk" the transfers based on the pad count used, then I feel that logic can be implemented within the controller driver, and should not leak out into the interface that MSPI devices use.

swift-tk replied:
The way I see it is that the list of settings would explode and eventually cover everything in `mspi_dev_cfg`.
You do realize that register access may only happened at lower frequencies? Then should we include frequency into `mspi_xfer`? An example device is ATXP032, the memory access frequencies can be higher than register access.
What we need is an API that balances between all these different controller and devices. Surely I value your input, but I also thinks that **we need to consider real world usage**. In general, for memory devices, you would have 90% of the memory access during run-time. For PSRAM, that number can go up to 99% as you would only need it for initialization and sleep.  For other SPI devices, such quirk back and forth with the io-mode is even rare as these quirks are really the result of memory manufaturers trying to bump memory performance. 

As I indicated in my other comment, the boundraies of APIs become blury when you have different APIs that could do the same thing. I prefer that a lower level API is clear and concise.

swift-tk replied:
To be honst, I see the list of settings in `mspi_xfer` starts to grow uncontrollably and we need discuss it with a base line for reference, something like JESD251C.

danieldegrasse commented on include/zephyr/drivers/mspi.h:
```diff
@@ -0,0 +1,791 @@
+/**
+ * @brief MSPI controller XIP configuration
+ */
+struct mspi_xip_cfg {
```
General question here- how do you envision supporting memory mapped access to MSPI peripherals? This is often the most efficient (and programmatically simple) method to configure XIP or interfacing with something like PSRAM.

RichardSWheatley replied:
I will leave the rest for Swoft as he is not on my time zone but he has written a full Ambiq MSPI driver, two intermediate chipset drivers for atxp032 and aps6404l and 2 samples for memc and flash.

Those will eventually be published here after MSPI is merged but maybe it would help the group if we were allowed to display them somewhere.

swift-tk replied:
> General question here- how do you envision supporting memory mapped access to MSPI peripherals? This is often the most efficient (and programmatically simple) method to configure XIP or interfacing with something like PSRAM.

I think @RichardSWheatley pretty much answered your question here. We do have a psram device driver planned for release in this PR.

danieldegrasse replied:
One other thing I would look at- we currently have a [FLEXSPI specific API](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/memc/memc_mcux_flexspi.h) within the tree that we use to support several PSRAM and flash devices. I don't think this API is a good candidate for a generic MSPI API, but it might be something to be aware of it you are interested in seeing how we are currently handling PSRAM

swift-tk replied:
> One other thing I would look at- we currently have a [FLEXSPI specific API](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/memc/memc_mcux_flexspi.h) within the tree that we use to support several PSRAM and flash devices. I don't think this API is a good candidate for a generic MSPI API, but it might be something to be aware of it you are interested in seeing how we are currently handling PSRAM

You should know that the FLEXSPI was the first thing I looked at when I moduled this MSPI API.

erwango replied:
> Those will eventually be published here after MSPI is merged but maybe it would help the group if we were allowed to display them somewhere.

This could indeed be useful to have a concrete usage example available to help converging on the API. This can be done as a side draft PR, which include current PR, and proposed as reference.

swift-tk replied:
I have one here if that helps you. https://github.com/AmbiqMicro/ambiqzephyr/pull/13

danieldegrasse commented on include/zephyr/drivers/mspi.h:
```diff
@@ -0,0 +1,791 @@
+ * @brief MSPI operational mode
+ */
+enum mspi_op_mode {
+	MSPI_OP_MODE_MASTER         = 0,
```
I'm confused when an MSPI device would operate in slave mode. Maybe I'm not understanding the design of this API correctly, but it seems like devices using the MSPI API would then implement other driver classes. For example,

- flash chips would implement the flash API using the MSPI API
- PSRAM chips would configure memory mapped access during init, and be implemented as MEMC drivers (no user facing API, although perhaps one is needed to get the base address of the MEMC device)
- other devices (displays, cameras, etc) would implement their relevant driver class

swift-tk replied:
Yes, your are correct. For slave mode, I do need someone to fill the void but I'd envision it also uses the proposed the API since the controller essentially act as a device itself and the devicetree structure still stands. The `mspi_config` would govern hardware specific and platform specific hardware settings. The `mspi_dev_config` would govern the device specific hardware settings (with respect to master).

For incomming transactions from master, it is essential to register a callback using mspi_register_callback so that the upper layer may be notified.

danieldegrasse replied:
Ok, so slave mode would be for a case where the MCU is acting as a receiver of QSPI/OSPI commands from another device? I think I understand the use case a bit better now- are you aware of any hardware that actually would use this? I suppose some FPGAs might present themselves as a slave on a QSPI bus, I'm just not clear what the software interface would look like for a peripheral that supports acting as a QSPI slave

swift-tk replied:
Like Communication between MCU and FPGA or between two MCUs. I think these are quite flexible in terms of settings I defined in the header file and would be for specific purposes not something you could just buy from market.

danieldegrasse commented on include/zephyr/drivers/mspi.h:
```diff
@@ -0,0 +1,791 @@
+/**
+ * @brief MSPI I/O mode capabilities
+ */
+enum mspi_io_mode {
```
Mentioned this below as well, but I don't think setting I/O mode this way makes sense. There are going to be cases where some commands for the device can run in 1-4-4 mode, and other commands are limited to 1-1-4 mode (or 1-1-1) mode for example. I think we need to set the *number of pads* each time we issue a SPI transfer

swift-tk replied:
Please see my explanation in https://github.com/zephyrproject-rtos/zephyr/pull/71769#discussion_r1596177507. 
When the mode of operation changes, it may be required to also reconfigure the device by sending commands, so calling `mspi_dev_config` before hand should be most appropriate.

RichardSWheatley commented:
@danieldegrasse I should have posted my replies in general instead of individual comments.

You can change many parameters on the fly from the chipset driver, prior to sending the command or data as each external chipset will have different configurations.

I am going to suggest we post some code here and hopefully that will help clear things up.

swift-tk commented:
For those that want to take a pi

swift-tk commented:
I clicked the wrong button…
For those that want to take a quick peek on the real hardware implementation please see here https://github.com/AmbiqMicro/ambiqzephyr/pull/13

RichardSWheatley commented:
Glad to see the latest commits. Including the Ambiq driver and the to chip drivers.

I think this should help clarify some questions and help move this ticket towards completion.

Is it possible to possibly discuss this in the next Architecture meeting to ensure everyone is on the same page?

erwango commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,198 @@
+==========
+With :c:struct:`mspi_dev_id` defined as collection of the device index and CE GPIO from
+device tree, the API supports multiple devices on the same controller instance.
+It is up to the controller driver implementation whether to support device switching
+either by software or by hardware. If by software, the switching should be performed
+in call to :c:struct:`mspi_dev_cfg`.
```
The controller driver implementation may or may not support device switching,
which can be performed either by software or by hardware. If the switching is handled
by software, it should be performed in a call to :c:struct:`mspi_dev_cfg`.

erwango commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,198 @@
+   };
+
+User should specify target operating parameters in the DTS such as mspi-max-frequency,
+mspi-io-mode and mspi-data-rate even though they may subject to change during run-time.
```
```suggestion
mspi-io-mode and mspi-data-rate even though they may be subject to change during run-time.
```

swift-tk commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,198 @@
+of each async/sync transfer if the callback had been registered using
+:c:func:`mspi_register_callback`. Or not to trigger any callback at all with
+:c:enum:`MSPI_BUS_NO_CB` even if the callbacks are already registered.
+In which case that a controller supports hardware command queue, user could take the full
+advantage of it in terms of performance if scatter IO and callback management are supported.
```
Please help me rephase this. What I meant is this:
user could take the full advantage of the hardware command queue in terms of performance if scattter IO and callback management are supported by the driver implementation under the assumption that the controller hardware supports hardware command queue.

swift-tk replied:
Thanks. :)

swift-tk committed:
tests: mspi: Add apollo3p_evb board to api test

Add the apollo3p_evb board to test the MSPI controller implementation.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk commented:
Hi all, 
Does anyone have further comments? If would like to discuss it, I'm available on discord or the architecture meeting.
There is one last and I would to direct your attention to the return values of the API that I have listed in the description. 
If everthing looks ok, I would like to move forward. It may be the time for the reviewers start to looked at the affected areas such as HW emulation, MEMC and FLASH. 

emulation: test:  @aaronemassey @jeremybettis @yperess @tristan-google @alevkoy @asemjonovs
flash: @de-nordic 
MEMC: ?
Devicetree/Devicetree Bindings: @galak @decsny 

Thanks!!

AlessandroLuo commented:
LGTM!

aaronyegx commented:
Hi @carlescufi @erwango @danieldegrasse , could you take another look and see if any other discussion or changes are needed? Thank you.

aaronyegx reviewed (dismissed):
MSPI is becoming more and more generic and it will be good to integrate the MSPI module to Zephyr for others to develop the driver for the vendor chips.
The Ambiq files look good to me. Looking forward more feedback about the architecture design and other affected files.

RichardSWheatley commented:
LGTM

FRASTM commented:
That's an interesting abstraction especially for the the multi-SPI NOR flash memories.
In some situation, the driver is getting the flash parameters from by the SFDP table of device memory itself, 
--> could values from the SFDP conflict with the values hold by the DTS of the device ?

carlescufi commented:
> Hi @carlescufi @erwango @danieldegrasse , could you take another look and see if any other discussion or changes are needed? Thank you.

I defer the Nordic review to @anangl. It would be great to get additional reviews from @erwango and @danieldegrasse

swift-tk commented:
> That's an interesting abstraction especially for the the multi-SPI NOR flash memories.
> In some situation, the driver is getting the flash parameters from by the SFDP table of device memory itself,
> --> could values from the SFDP conflict with the values hold by the DTS of the device ?

The short answer is it could be conflicting to the SFDP. If there is a conflict, then there will probably be problems during run time. Notice that the values hold by device DTS are brought into the device driver, so the device initialization can be optimized by including a SFDP read operation and then check against the target settings in the DTS to catch incompatibility. The base settings for reading the SFDP should be hardcoded in the device driver and match the factory default settings of the device after power up. In my case, the flash atxp032 driver, there is a const serial_cfg that stores the default settings. In which case that a flash device powers up from their OTP state, one could introduce another otp_cfg variable.

swift-tk commented:
@anangl @FRASTM @GeorgeCGV @danieldegrasse Any additional comments?
If not, could we move forward? @carlescufi

kartben commented on include/zephyr/drivers/mspi.h:
```diff
@@ -0,0 +1,808 @@
+ * @brief MSPI operational mode
+ */
+enum mspi_op_mode {
+	MSPI_OP_MODE_MASTER         = 0,
+	MSPI_OP_MODE_SLAVE          = 1,
```
introducing a new API that is using non inclusive language is a no-go IMO
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-2-inclusive-language

RichardSWheatley replied:
The SPI standard was written as MASTER/SLAVE.
The new terms are typically main/sub.

Where and how in Zephyr would we describe to a user that MAIN means MASTER and SUB means SLAVE and how would you do that without using the word, master and slave?

kartben replied:
> The new terms are typically main/sub.

So I guess

```suggestion
	MSPI_OP_MODE_MAIN         = 0,
	MSPI_OP_MODE_SUB          = 1,
```

OSHWA seems to be using controller/peripheral, fwiw The new terms are typically main/sub.

btw please don't mark comments as resolved as the discussion is ongoing :)

dleach02 replied:
Actually, the[ Revision 1.5 from May 22](https://downloadmirror.intel.com/27055/327432%20espi_base_specification%20R1-5.pdf) changed the language to Controller/Target:


• Updated the terms “master/slave” to “controller/target”
throughout the specification to align with the use of
inclusive language

But I think we should be able to follow this up with a quick PR to change everything over.

RichardSWheatley replied:
Yes, I just saw your comments David!
Thanks for posting that.

Should we also change to CITO and COTI?
That would make sense, correct?

RichardSWheatley replied:
> > The new terms are typically main/sub.
> 
> 
> 
> OSHWA seems to be using controller/peripheral, fwiw https://www.oshwa.org/a-resolution-to-redefine-spi-signal-names/
> 
> 
> 
> btw please don't mark comments as resolved as the discussion is ongoing :)

Sorry, I am using Biq bucket on my phone at the doctor's office while getting a blood draw.

I should put the phone down.

dleach02 commented on doc/hardware/peripherals/mspi.rst:
```diff
@@ -0,0 +1,199 @@
+``mspi-io-mode`` and ``mspi-data-rate`` even though they may subject to change during runtime.
+It should represent the typical configuration of the device during normal operations.
+
+Multi-slave
```
Will need to change this to align terminology with the SPI specification. So likely "multi-target".

From the SPI specification:

Updated the terms “master/slave” to “controller/target”
throughout the specification to align with the use of
inclusive language.

RichardSWheatley replied:
Thank you! I literally just had a question I post as a reply in another comment.

So controller/target.

What about COPI, PICO for "peripheral and controller", or COTI for "controller" and "target"?

dleach02 commented on dts/bindings/mspi/mspi-controller.yaml:
```diff
@@ -0,0 +1,94 @@
+
+  op-mode:
+    type: string
+    enum:
```
same comment, will need to change to controller/target in a follow-up PR

dleach02 commented on dts/bindings/mspi/mspi-device.yaml:
```diff
@@ -0,0 +1,212 @@
+  mspi-hold-ce:
+    type: boolean
+    description: |
+      In some cases, it is necessary for the master to manage
```
"controller"

RichardSWheatley replied:
No need to comment every single instance.
I believe the understanding is to switch to controller and target. 

Please see my comment on your other message. .

swift-tk commented:
@kartben @dleach02 I agree with nashif. To be fair, the SPI API has not adapted that and the terms in MSPI are taken from SPI. If it is not blocking and not absolutely necessary, then I prefer not to change anything at this time.

RichardSWheatley commented:
> @kartben @dleach02 I agree with nashif. To be fair, the SPI API has not adapted that and the terms in MSPI are taken from SPI. If it is not blocking and not absolutely necessary, then I prefer not to change anything at this time.

@swift-tk one of @dleach02 last comments was that this could be done in a follow-up PR.

@dleach02 ia that ok?

dleach02 commented:
> > @kartben @dleach02 I agree with nashif. To be fair, the SPI API has not adapted that and the terms in MSPI are taken from SPI. If it is not blocking and not absolutely necessary, then I prefer not to change anything at this time.
> 
> @swift-tk one of @dleach02 last comments was that this could be done in a follow-up PR.
> 
> @dleach02 ia that ok?

we are discussing that in discord. The suggestion was that it was okay but we would enter a blocking bug against the API to be addressed during the v3.7 stabilization period.

dleach02 reviewed (dismissed):
@kartben, they updated the PR to make the language translations.

danieldegrasse commented:
> Hi @carlescufi @erwango @danieldegrasse , could you take another look and see if any other discussion or changes are needed? Thank you.

Taking another look now- apologies, I was not aware we were trying to cut this in before 3.7

RichardSWheatley commented:
> > Hi @carlescufi @erwango @danieldegrasse , could you take another look and see if any other discussion or changes are needed? Thank you.
> 
> 
> 
> Taking another look now- apologies, I was not aware we were trying to cut this in before 3.7

Where can we look for dates on 3.7 LTS fr ze and release?

swift-tk commented:
> > > Hi @carlescufi @erwango @danieldegrasse , could you take another look and see if any other discussion or changes are needed? Thank you.
> > 
> > 
> > Taking another look now- apologies, I was not aware we were trying to cut this in before 3.7
> 
> Where can we look for dates on 3.7 LTS fr ze and release?

https://github.com/zephyrproject-rtos/zephyr/wiki/Release-Management

danieldegrasse commented on dts/bindings/mtd/mspi-atxp032.yaml:
```diff
@@ -0,0 +1,8 @@
+
+compatible: "mspi-atxp032"
+
+include: ambiq,mspi-device.yaml
```
Shouldn't this be `mspi-device.yaml`?

swift-tk replied:
Yes, it got left out as an unresolved issue. The ATXP032 flash device driver is intended to be a universal driver, but I needed to pass some properties that is unique to Ambiq from DTS. Now that I think of it, I should be able to add the ambiq,mspi-device compat to the DTS instead of including it here.

danieldegrasse reviewed (commented):
@swift-tk I've taken another look, and I think I'm alright with the changes. There may need to be some API changes in the future to support other controllers, but as far as I can tell this would work fine with the FLEXSPI. I am ok with merging this, provided you are ok with the *possibility* of some API changes after 3.7

A few questions I had:
- The API seems to be designed to transition the device into QSPI or OSPI mode, and then send all commands in that mode. I guess for devices that do not support QSPI/OSPI, we would set them to 1-4-4 or 1-8-8 mode?
- Are there devices you are aware of that use something like 1-4-4 mode for flash reads and 1-2-2 mode for flash writes? From my understanding, that would result in frequent reconfiguration with this API. I suppose the idea in that case would be that read commands would be handled via the `mspi_dev_cfg`, since that configuration is static and does not need to change?
- When using 1-4-4 or 1-8-8 mode, is my understanding correct that the first byte of a transfer would be sent using 1 pad, and future bytes of the same transfer would use 4/8 pads?
- How will this API work with SFDP? It seems like it would function by probing all data in serial mode, then reconfiguring the device based on what was discovered, is that correct?

danieldegrasse commented:
Running `tests/subsys/settings/functional/file` on `native_sim//64` seems to also fail on `main`

danieldegrasse reviewed (approved):
Restarted failed CI test

swift-tk commented:
> The API seems to be designed to transition the device into QSPI or OSPI mode, and then send all commands in that mode. I guess for devices that do not support QSPI/OSPI, we would set them to 1-4-4 or 1-8-8 mode?
What mode to be set should be based on the device datasheet, i.e. transfer waveforms. And yes, the transcieve API is designed to send everything in one particular mode.

> Are there devices you are aware of that use something like 1-4-4 mode for flash reads and 1-2-2 mode for flash writes? From my understanding, that would result in frequent reconfiguration with this API. I suppose the idea in that case would be that read commands would be handled via the `mspi_dev_config`, since that configuration is static and does not need to change?
I'm aware of one particular device that could only write memory with 1-1-4 mode and able to read memory using 1-1-4 and 1-4-4. But I think this is rare and poorly designed hardware in attempt to bump read speed. In this case, whether it result in frequent reconfiguration or not will depend on the usage. i.e. the percentages of write and read memory commands. If it is a flash device we are talking about, write may already be very costly with erase and page layout are involved. Everything will still need to go through `mspi_dev_config`, as transceive API is not aware of io-mode and etc. If case by case analysis that the configuration becomes very frequent, then I suggest to use a software queue like RTIO to manage calls to MSPI API to bascially pipelining the requests. 
I understand that the hardware command queue are powerful and capable of handling switching every settable things. But I'd prefer that the controller driver in general shouldn't need to manage the internal states of its device as I think it is better handled at device drivers. If we attempt to do that, we would probably end of having similar spi_cfg with all configurable parameters when calling transceive API. It would become very unfriendly to use as the number of parameters would be quite large for MSPI.

> When using 1-4-4 or 1-8-8 mode, is my understanding correct that the first byte of a transfer would be sent using 1 pad, and future bytes of the same transfer would use 4/8 pads?
1-4-4 correspond to sending command with 1 pad, sending address with 4 pad and sending data with 4 pad.

> How will this API work with SFDP? It seems like it would function by probing all data in serial mode, then reconfiguring the device based on what was discovered, is that correct?
For that particular flash ATXP032 driver I did, if I understand things correctly in the datasheet, the SFDP can only be read in serial mode. The device also always power up into serial mode. But this may not be true for other devices.
Constructing the SFDP API would be device driver specific and it would just follow the same flow as with write and read memory commands. If the device need to be switched to certain io mode or certain state in order to read SFDP, then do the switch with `mspi_dev_config`, once read is finished, it should switch itself back to normal operation mode.

swift-tk commented:
I thank everyone for their effort to make the merge happen!! But it is not done, there are rooms to improve for sure. I'm happy to further improve the API or review any PR to improve it.
Cheers!!
